compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp

```diff
@@ -210,11 +210,108 @@ struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
   }
 };
 
+// Truncates input operand if it is a constant and is out of range of target
+// type.
+Value truncateOutOfRangeValue(Value operand,
+                              ConversionPatternRewriter &rewriter,
+                              TypeConverter *typeConverter) {
+  if (!operand.getDefiningOp()) {
+    return operand;
+  }
+  auto oldType = operand.getType();
+  auto newType = typeConverter->convertType(oldType);
+  if (oldType == newType) {
+    return operand;
+  }
+  auto constOp = dyn_cast<arith::ConstantOp>(operand.getDefiningOp());
+  if (!constOp) {
+    return operand;
+  }
+  auto intAttr = constOp->getAttr("value").dyn_cast<IntegerAttr>();
+  APInt value = intAttr.getValue();
+  APInt newValue = value.truncSSat(newType.getIntOrFloatBitWidth());
+  if (newValue == value) {
+    return operand;
+  }
+  Value truncatedConst = rewriter.create<arith::ConstantOp>(
+      constOp->getLoc(), IntegerAttr::get(oldType, newValue));
+  return truncatedConst;
+}
+
+// Utility function to check that op is signed or not.
+bool isSignedOp(Operation *op) {
+  if (isa<arith::CeilDivSIOp>(op) || isa<arith::DivSIOp>(op) ||
+      isa<arith::ExtSIOp>(op) || isa<arith::FloorDivSIOp>(op) ||
+      isa<arith::MaxSIOp>(op) || isa<arith::RemSIOp>(op) ||
+      isa<arith::SIToFPOp>(op) || isa<arith::ShRSIOp>(op)) {
+    return true;
+  }
+  if (isa<arith::CmpIOp>(op)) {
+    bool isSigned = op->getAttr("predicate")
+                                .dyn_cast<mlir::arith::CmpIPredicateAttr>()
+                                .getValue() < mlir::arith::CmpIPredicate::ult
+                        ? true
+                        : false;
+    return isSigned;
+  }
+  if (isa<arith::SelectOp>(op)) {
+    for (auto childOp : op->getUsers()) {
+      bool childOpSign = isSignedOp(childOp);
+      if (childOpSign) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+struct ConvertSignSensitiveArithCmpIOp
+    : public OpConversionPattern<arith::CmpIOp> {
+  using OpConversionPattern<arith::CmpIOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      arith::CmpIOp op, typename arith::CmpIOp::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto cmpPredicate = op->getAttr("predicate")
+                            .dyn_cast<mlir::arith::CmpIPredicateAttr>()
+                            .getValue();
+    bool isSigned = isSignedOp(op);
+    if (!isSigned) return failure();
+    Value lhs = truncateOutOfRangeValue(op.getOperand(0), rewriter,
+                                        this->getTypeConverter());
+    Value rhs = truncateOutOfRangeValue(op.getOperand(1), rewriter,
+                                        this->getTypeConverter());
+    if (lhs == op.getOperand(0) && rhs == op.getOperand(1)) return failure();
+    rewriter.replaceOpWithNewOp<arith::CmpIOp>(op, cmpPredicate, lhs, rhs);
+    return success();
+  }
+};
+
+struct ConvertSignSensitiveArithSelectOp
```

Comment:
Do we need two separate classes here? What about a single pattern matching everything to handle both? Something like:

```c++
struct ConvertSignSensitiveArithOp: public OpConversionPattern {
    ConvertSignSensitiveArithOp(MLIRContext *context, TypeConverter &typeConverter)
      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), /*patternBenefit=*/1, context) {}

  LogicalResult matchAndRewrite(
      Operation *op, ArrayRef<Value> operands,
      ConversionPatternRewriter &rewriter) const override {
        if (!isSignedOp(op)) return failure();
        SmallVector<Value, 4> newOperands;
        bool truncated = false;
        for (Value oldOperand: op->getOperands()) {
          Value newOperand = truncateOutOfRangeValue(...);
          newOperands.push_back(newOperand);
          truncated |= (oldOperand != newOperand);
        }
        if (!truncated) return failure();
        OperationState state(op->getLoc(), op->getName().getStringRef(), newOperands,
                                             op->getResultTypes(), op->getAttrs());
        Operation *newOp = rewriter.create(state);
        rewriter.replaceOp(op, newOp->getResults());
        return success();
      }
};
```

Reply:
Done! thanks :)